log: print every script verification state change

l0rinc opened the pull request:
### Summary

Users can encounter cases where script checks are unexpectedly enabled (e.g. after reindex, or when `assumevalid`/`minimumchainwork` gates fail). Without an explicit line, they must infer state from the absence of a message, which is incomplete and error-prone. 
The existing "Assuming ancestors of block …" line does not reliably indicate whether script checks are actually enabled, which makes debugging/benchmarking confusing.

### What this changes

We make the initial **script-verification** state explicit and log **why** checks are enabled to avoid confusion.
* Always log the first script-verification state on startup, **before** the first `UpdateTip`.
* Flatten the nested `assumevalid` conditionals into a linear gating sequence for readability.
* Extend the functional test to assert the old behavior with the new reason strings.

This is a **logging-only** test change it shouldn't change any other behavior.

### Example output

The state (with reason) is logged at startup and whenever the reason changes, e.g.:

* `Disabling script verification at block #904336 (000000000000000000014106b2082b1a18aaf3091e8b337c6fed110db8c56620).`
* `Enabling script verification at block #912527 (000000000000000000010bb6aa3ecabd7d41738463b6c6621776c2e40dbe738a): block too recent relative to best header.`
* `Enabling script verification at block #912684 (00000000000000000001375cf7b90b2b86e559d05ed92ca764d376702ead3858): block height above assumevalid height.`

------

Follow-up to https://github.com/bitcoin/bitcoin/pull/32975#discussion_r2329269037

l0rinc committed:
log: separate script verification reasons

Replace `fScriptChecks` with `script_check_reason` and log the precise reason when checks are enabled; log a plain "Disabling" when they are skipped.
Adjust the functional test to assert the new reason strings.

Co-authored-by: w0xlt <woltx@protonmail.com>
Co-authored-by: Eunovo <eunovo9@gmail.com>
Co-authored-by: Hodlinator <172445034+hodlinator@users.noreply.github.com>
Co-authored-by: TheCharlatan <seb.kung@gmail.com>
Co-authored-by: Andrew Toth <andrewstoth@gmail.com>

l0rinc committed:
validation: log initial script verification state

Replaced `atomic<bool>` with `std::optional<bool>` (logged once on first observation). Safe because `ConnectBlock` holds `cs_main`.\
After this change, the state is logged before the very first `UpdateTip` line.

Co-authored-by: Eunovo <eunovo9@gmail.com>
Co-authored-by: Hodlinator <172445034+hodlinator@users.noreply.github.com>
Co-authored-by: w0xlt <woltx@protonmail.com>

yuvicc reviewed (commented):
Concept ACK. 

The approach of switching to `std::optional<bool>` here avoids the overhead of atomic operations relying on `cs_main` for synchronization makes sense. 

<details> <summary>Initial script verification state is always logged without unnecessary complexity.</summary>

```bash
2025-09-08T07:43:58Z Disabling signature validations at block #16893 (00000000113c355415c3417c734a45b62790b91b41ff08028580254c584f5ea8).
```

</details>

l0rinc commented:
For more context, even just restarting the node that was reindexed before (and is at around 200k blocks) with `-assumevalid=0000000000000000000087564caa77e7b3f29d0464256c04d5539e43663f8698` (block 914005), it still starts with signature verification enabled - using 70% of CPU when it should have been disabled in the first place, see:
<img width="1000" alt="image" src="https://github.com/user-attachments/assets/e6094caf-c0d3-40dd-8edd-92cbe76c3d81" />
This PR makes it at least obvious that something's off.
https://github.com/bitcoin/bitcoin/pull/31615#issuecomment-3273275956 doesn't seem to fix the issue, but that's orthogonal with at least making it obvious via the logs.

Eunovo commented on src/validation.h:
```diff
@@ -560,7 +560,7 @@ class Chainstate
     mutable const CBlockIndex* m_cached_snapshot_base GUARDED_BY(::cs_main){nullptr};
 
-    std::atomic_bool m_prev_script_checks_logged{true};
+    std::optional<bool> m_prev_script_checks_logged{};
```
https://github.com/bitcoin/bitcoin/pull/33336/commits/64ba2821b5a55aa6d51fbb7a16470d4f7c41c92e:
I believe you now need to add `GUARDED_BY(::cs_main)` here to use Clang's Thread Safety Analysis

l0rinc replied:
Excellent, added you as a coathor, thanks.

hodlinator reviewed (approved):
ACK 8cb75ee0a63880b55ac695fe884748aa7d604c38

While it is unsettling that we don't fully understand what causes script checking to occur prior to the assumevalid height (https://github.com/bitcoin/bitcoin/pull/32975#discussion_r2329269037) - having an initial "`Enabling signature validations at block #1 ...`" in `-assumevalid=0` cases as per this PR seems fine regardless.

(Empty `std::optional` does not match `bool fScriptChecks` until set).

https://github.com/bitcoin/bitcoin/blob/8cb75ee0a63880b55ac695fe884748aa7d604c38/src/validation.cpp#L2566-L2569

(Memory behavior when locking/unlocking a mutex (`cs_main`) does AFAIK imply `atomic_` on master was never necessary).

sedited reviewed (approved):
ACK 8cb75ee0a63880b55ac695fe884748aa7d604c38

The log message is a bit clunky though. Since it now appears on startup I think it would be nice if it also mentioned that this just reflects assumevalid.

l0rinc commented on test/functional/feature_assumevalid.py:
```diff
@@ -146,7 +146,8 @@ def run_test(self):
-        self.send_blocks_until_disconnected(p2p0)
+        with self.nodes[0].assert_debug_log(expected_msgs=['Enabling signature validations at block #1']):
+            self.send_blocks_until_disconnected(p2p0)
         self.wait_until(lambda: self.nodes[0].getblockcount() >= COINBASE_MATURITY + 1)
```
This seems less racy, indeed, updated, added you as coauthor, thanks

w0xlt reviewed (commented):
Approach ACK

Using `std::optional<bool>` introduces a third state (unset), which guarantees exactly one initial log message—regardless of whether the first real value is true or false.

For reviewers, here’s a simple way to verify the behavior change:

```
mkdir temp_data_dir
./build/bin/bitcoind -regtest -datadir=temp_data_dir -daemon
./build/bin/bitcoin-cli -regtest -datadir=temp_data_dir createwallet w
./build/bin/bitcoin-cli -regtest -datadir=temp_data_dir getnewaddress
./build/bin/bitcoin-cli -regtest -datadir=temp_data_dir generatetoaddress 1 <address>
```

With this PR, the logs will include:
```
Enabling signature validations at block #1
```
In master, this line does not appear.

Eunovo reviewed (commented):
ACK https://github.com/bitcoin/bitcoin/pull/33336/commits/39f41135c764c6c4705ce2cd36781ca8d43f0114:

Previously, users could only verify that script checks were enabled by confirming the absence of a "Disabling signature validations at block #1" message in the logs, which is a confusing negative confirmation pattern.
This PR adds explicit positive feedback by ensuring an "Enabling signature validations at block #1" message appears when script checks are enabled, making the system state clearer and easier to verify.

hodlinator reviewed (approved):
re-ACK 39f41135c764c6c4705ce2cd36781ca8d43f0114

Only made test less racy (https://github.com/bitcoin/bitcoin/pull/33336#discussion_r2337774261) since the first ACK.

---

> > After `reindex`, my node has ~841k headers, a lot lower than the current tip of ~914k.
> > The default `assumevalid` block (height 912,683) is in that missing range, [would say: is outside that range]
> 
> I think the behaviour here is probably fine -- it's a weak way of preventing you from triggering assumevalid behaviour on an invalid chain with substantially lower total work than the main chain. It might be nice to have a warning that your headers chain doesn't meet minchainwork though. It would probably be nicer to be able to obtain the missing headers though.
> 
> EDIT: probably fine, compared to the alternative of dropping the `best_header->nChainWork >= MinimumChainWork` check, and just relying on the assumevalid block being in the best_header's ancestry. That didn't seem to have much rationale when it was introduced afaics: [#9484 (comment)](https://github.com/bitcoin/bitcoin/pull/9484#issuecomment-272469033)

(Bottom of https://github.com/bitcoin/bitcoin/pull/9484/commits/e440ac7ef3b6f3ad1cd8fc7027cece40413202d9 is interesting, enforcing script validation within the last 2 weeks regardless of assumevalid settings. It's annoying that the current default assume valid block height of 912'683 is ~10 days ago, but shouldn't matter much).

I would hope this condition kicked in and filled in the missing headers from 841k to the `nMinimumChainWork` at 912'683 so that the best header could be updated and script validation could be turned off again until the last 2 weeks of blocks:

https://github.com/bitcoin/bitcoin/blob/39f41135c764c6c4705ce2cd36781ca8d43f0114/src/net_processing.cpp#L5526-L5539

Any clue to why missing headers are not being fetched / best header isn't updated?

w0xlt reviewed (commented):
ACK https://github.com/bitcoin/bitcoin/pull/33336/commits/39f41135c764c6c4705ce2cd36781ca8d43f0114

Simple test on mainnet:
```
./build/bin/bitcoind -datadir=new_data_dir/ -assumevalid=000000009b7262315dbf071787ad3656097b892abffd1f95a1a022f896f533fc -stopatheight=10
```

l0rinc committed:
test: add assumevalid scenarios scaffold

Increase the test to 6 nodes and add flows for baseline, deep anchor, and too-recent cases, plus scaffolding for off-best-header, not-in-assumevalid,
and reindex gates.
Assertions are minimal here; follow-ups add reason checks.

Co-authored-by: Hodlinator <172445034+hodlinator@users.noreply.github.com>

l0rinc committed:
refactor: untangle assumevalid decision branches

Flatten nested conditionals into a linear gating sequence for readability and precise logging. No functional change, TODOs are addressed in next commit

w0xlt commented:
Approach ACK.

> beneficial to extend this PR to show the reason for why script verification was enabled.

I agree.

> split the above suggestion into small focused commits

Commits could be squashed but but keeping them separate makes it clearer to review.

w0xlt commented on src/validation.cpp:
```diff
@@ -2423,34 +2439,41 @@ bool Chainstate::ConnectBlock(const CBlock& block, BlockValidationState& state,
+            assume_valid = CHECKED_OFF_BESTHEADER_PATH;
+        } else if (m_chainman.m_best_header->nChainWork < m_chainman.MinimumChainWork()) {
+            assume_valid = CHECKED_BELOW_MIN_CHAINWORK;
+        } else if (GetBlockProofEquivalentTime(*m_chainman.m_best_header, *pindex, *m_chainman.m_best_header, params.GetConsensus()) <= 60 * 60 * 24 * 7 * 2) {
```
Just to make clearer what  `60 * 60 * 24 * 7 * 2` means.

```cpp
constexpr int64_t TWO_WEEKS = 14 * 24 * 60 * 60;
if (GetBlockProofEquivalentTime(... ) <= TWO_WEEKS) { ... }
```

Or (not tested)

```cpp
#include <chrono>
using namespace std::chrono;

constexpr auto TWO_WEEKS = 14d; // 14 days

if (GetBlockProofEquivalentTime(...) <= duration_cast<seconds>(TWO_WEEKS).count()) {
    ...
}
```

w0xlt commented on src/validation.cpp:
```diff
@@ -106,6 +108,20 @@ const std::vector<std::string> CHECKLEVEL_DOC {
+    case CHECKED_HASH_NOT_IN_HEADERS:   return "assumevalid hash not in headers";
+    case CHECKED_NOT_UNDER_ASSUMEVALID: return "block not under assumevalid anchor";
+    case CHECKED_OFF_BESTHEADER_PATH:   return "block not on best header path";
+    case CHECKED_BELOW_MIN_CHAINWORK:   return "best header chainwork below nMinimumChainWork";
```
nit: logs shouldn’t expose internal variable names

```suggestion
    case CHECKED_BELOW_MIN_CHAINWORK:   return "best header chainwork below -minimumchainwork";
```

w0xlt reviewed (commented):
ACK https://github.com/bitcoin/bitcoin/pull/33336/commits/63dc937eea297332ec85c1abbcf3b94b8f74aa85 with non-blocking nits mentioned above.

Eunovo reviewed (commented):
Concept ACK https://github.com/bitcoin/bitcoin/pull/33336/commits/63dc937eea297332ec85c1abbcf3b94b8f74aa85
Left some comments.

l0rinc reviewed (commented):
Thanks for the suggestions, applied most of them

hodlinator commented on src/validation.cpp:
```diff
@@ -106,6 +108,20 @@ const std::vector<std::string> CHECKLEVEL_DOC {
+    case CHECKED_OFF_BESTHEADER_PATH:   return "block not on best header path";
+    case CHECKED_BELOW_MIN_CHAINWORK:   return "best header chainwork below nMinimumChainWork";
+    case CHECKED_NOT_BURIED_ENOUGH:     return "too recent relative to best header";
+    default:                            std::abort();
```
Please consider triggering a compiler warning instead.
```
₿ git grep "// no default case, so the compiler can warn about missing cases" |wc -l
70
```
https://github.com/bitcoin/bitcoin/blob/9b04e4f96200daf7516d1b8b6b0dfc4c077837e8/src/deploymentinfo.cpp#L36-L37

w0xlt reviewed (commented):
reACK https://github.com/bitcoin/bitcoin/pull/33336/commits/9b04e4f96200daf7516d1b8b6b0dfc4c077837e8

hodlinator reviewed (commented):
Reviewed 9b04e4f96200daf7516d1b8b6b0dfc4c077837e8

hodlinator commented on src/validation.cpp:
```diff
@@ -2423,34 +2438,42 @@ bool Chainstate::ConnectBlock(const CBlock& block, BlockValidationState& state,
+            assume_valid = CHECKED_HASH_NOT_IN_HEADERS;
+        } else if (it->second.GetAncestor(pindex->nHeight) != pindex) {
+            assume_valid = CHECKED_NOT_UNDER_ASSUMEVALID;
+        } else if (m_chainman.m_best_header->GetAncestor(pindex->nHeight) != pindex) {
+            assume_valid = CHECKED_OFF_BESTHEADER_PATH;
```
Found a way to test 2):
```python
        # nodes[4]
        self.start_node(4, extra_args=["-assumevalid=" + block102.hash_hex])
        second_chain_tip = int(self.nodes[4].getbestblockhash(), 16)
        second_chain_time = self.nodes[4].getblock(self.nodes[4].getbestblockhash())['time'] + 1
        second_chain_height = self.nodes[4].getblock(self.nodes[4].getbestblockhash())['height'] + 1

        second_chain = []
        for _ in range(2100):
            block = create_block(second_chain_tip, create_coinbase(second_chain_height), second_chain_time)
            block.solve()
            second_chain.append(block)
            second_chain_tip = block.hash_int
            second_chain_time += 1
            second_chain_height += 1

        p2p4 = self.nodes[4].add_p2p_connection(BaseNode())
        p2p4.send_header_for_blocks(second_chain[0:2000])
        p2p4.send_header_for_blocks(second_chain[2000:])
        p2p4.send_header_for_blocks(self.blocks[0:103])

        with self.nodes[4].assert_debug_log(expected_msgs=[f'Enabling signature validations at block #1 ({block_1_hash}): block not on best header chain']):  # AssumeValid::CHECKED_OFF_BESTHEADER_PATH
            # Send all blocks to node1. All blocks will be accepted.
            p2p4.send_without_ping(msg_block(self.blocks[0]))
            self.wait_until(lambda: self.nodes[2].getblockcount() >= 1)
```

l0rinc committed:
log: reword `signature validations` to `script verification` in `assumevalid` log

Even though not all script verification is turned off currently (e.g. we're still doing the cheaper sigop counts), this naming is more consistent with other usages.

l0rinc reviewed (commented):
Thanks for the reviews, I have simplified the commits, removed the enum and replaced it with direct reason string, added an `off best‑header path` (thanks to @hodlinator), reworded the log to say "script verification" instead of "signature validations" to be in alignment with the "Script verification uses ... additional threads" and reverted the `fScriptChecks` and `m_prev_script_checks_logged` symbols.

There's a remaining suggestion to have friendlier reason in case of:
```
2025-09-16T21:00:03Z Enabling script verification at block #915016 (000000000000000000001a6f6115c32c340a2159b30f9ffd632f63424aa563f9): block not part of assumevalid chain.
```
to separate when the block has the assumevalid as an ancestor. I found that a bit riskier change than the rest, I think we should do that in a follow-up instead.

l0rinc commented on src/validation.cpp:
```diff
@@ -2423,7 +2423,7 @@ bool Chainstate::ConnectBlock(const CBlock& block, BlockValidationState& state,
     }
 
-    bool fScriptChecks = true;
+    auto assume_valid{true};
```
Ii have reverted the naming

l0rinc commented on src/validation.cpp:
```diff
@@ -2564,7 +2574,14 @@ bool Chainstate::ConnectBlock(const CBlock& block, BlockValidationState& state,
-        LogInfo("%s signature validations at block #%d (%s).", fScriptChecks ? "Enabling" : "Disabling", pindex->nHeight, block_hash.ToString());
+        if (fScriptChecks) {
+            Assume(!script_check_reason.empty());
+            LogInfo("Enabling script verification at block #%d (%s): %s.",
+                    pindex->nHeight, block_hash.ToString(), script_check_reason);
+        } else {
+            LogInfo("Disabling script verification at block #%d (%s).",
+                    pindex->nHeight, block_hash.ToString());
```
Currently we're only logging this if there's a single one, i.e. `GetRole() == ChainstateRole::NORMAL`

hodlinator commented on src/validation.cpp:
```diff
@@ -2423,35 +2423,45 @@ bool Chainstate::ConnectBlock(const CBlock& block, BlockValidationState& state,
     }
 
-    bool fScriptChecks = true;
+    auto fScriptChecks{true};
+    std::string script_check_reason;
```
nit: I'd prefer avoiding having this dual state. Here are 2 variations which do that:

Preserving behavior:

```diff
diff --git a/src/validation.cpp b/src/validation.cpp
index 7e82c3bbad..008bfe4d71 100644
--- a/src/validation.cpp
+++ b/src/validation.cpp
@@ -2423,8 +2423,7 @@ bool Chainstate::ConnectBlock(const CBlock& block, BlockValidationState& state,
         return true;
     }
 
-    auto fScriptChecks{true};
-    std::string script_check_reason;
+    const char* script_check_reason{nullptr};
     if (!m_chainman.AssumedValidBlock().IsNull()) {
         constexpr int64_t TWO_WEEKS_IN_SECONDS{60 * 60 * 24 * 7 * 2};
         // We've been configured with the hash of a block which has been externally verified to have a valid history.
@@ -2458,7 +2457,6 @@ bool Chainstate::ConnectBlock(const CBlock& block, BlockValidationState& state,
             //  artificially set the default assumed verified block further back.
             // The test against the minimum chain work prevents the skipping when denied access to any chain at
             //  least as good as the expected chain.
-            fScriptChecks = false;
         }
     } else {
         script_check_reason = "assumevalid=0 (always verify)";
@@ -2573,16 +2571,15 @@ bool Chainstate::ConnectBlock(const CBlock& block, BlockValidationState& state,
              Ticks<SecondsDouble>(m_chainman.time_forks),
              Ticks<MillisecondsDouble>(m_chainman.time_forks) / m_chainman.num_blocks_total);
 
-    if (fScriptChecks != m_prev_script_checks_logged && GetRole() == ChainstateRole::NORMAL) {
-        if (fScriptChecks) {
-            Assume(!script_check_reason.empty());
+    if ((script_check_reason != nullptr) != m_prev_script_checks_logged && GetRole() == ChainstateRole::NORMAL) {
+        if (script_check_reason) {
             LogInfo("Enabling script verification at block #%d (%s): %s.",
                     pindex->nHeight, block_hash.ToString(), script_check_reason);
         } else {
             LogInfo("Disabling script verification at block #%d (%s).",
                     pindex->nHeight, block_hash.ToString());
         }
-        m_prev_script_checks_logged = fScriptChecks;
+        m_prev_script_checks_logged = script_check_reason != nullptr;
     }
 
     CBlockUndo blockundo;
@@ -2593,7 +2590,7 @@ bool Chainstate::ConnectBlock(const CBlock& block, BlockValidationState& state,
     // doesn't invalidate pointers into the vector, and keep txsdata in scope
     // for as long as `control`.
     std::optional<CCheckQueueControl<CScriptCheck>> control;
-    if (auto& queue = m_chainman.GetCheckQueue(); queue.HasThreads() && fScriptChecks) control.emplace(queue);
+    if (auto& queue = m_chainman.GetCheckQueue(); queue.HasThreads() && script_check_reason != nullptr) control.emplace(queue);
 
     std::vector<PrecomputedTransactionData> txsdata(block.vtx.size());
 
@@ -2652,7 +2649,7 @@ bool Chainstate::ConnectBlock(const CBlock& block, BlockValidationState& state,
             break;
         }
 
-        if (!tx.IsCoinBase() && fScriptChecks)
+        if (!tx.IsCoinBase() && script_check_reason != nullptr)
         {
             bool fCacheResults = fJustCheck; /* Don't cache results if we're actually connecting blocks (still consult the cache, though) */
             bool tx_ok;
```

<details><summary>The below version will output more often, when switching reasons, not sure which behavior is more desirable (there is no risk of dangling pointers from using `const char*`, but it requires us to be vigilant going forward):</summary>

```diff
diff --git a/src/validation.cpp b/src/validation.cpp
index 7e82c3bbad..79ad71b64c 100644
--- a/src/validation.cpp
+++ b/src/validation.cpp
@@ -2423,8 +2423,7 @@ bool Chainstate::ConnectBlock(const CBlock& block, BlockValidationState& state,
         return true;
     }
 
-    auto fScriptChecks{true};
-    std::string script_check_reason;
+    const char* script_check_reason{nullptr};
     if (!m_chainman.AssumedValidBlock().IsNull()) {
         constexpr int64_t TWO_WEEKS_IN_SECONDS{60 * 60 * 24 * 7 * 2};
         // We've been configured with the hash of a block which has been externally verified to have a valid history.
@@ -2458,7 +2457,6 @@ bool Chainstate::ConnectBlock(const CBlock& block, BlockValidationState& state,
             //  artificially set the default assumed verified block further back.
             // The test against the minimum chain work prevents the skipping when denied access to any chain at
             //  least as good as the expected chain.
-            fScriptChecks = false;
         }
     } else {
         script_check_reason = "assumevalid=0 (always verify)";
@@ -2573,16 +2571,15 @@ bool Chainstate::ConnectBlock(const CBlock& block, BlockValidationState& state,
              Ticks<SecondsDouble>(m_chainman.time_forks),
              Ticks<MillisecondsDouble>(m_chainman.time_forks) / m_chainman.num_blocks_total);
 
-    if (fScriptChecks != m_prev_script_checks_logged && GetRole() == ChainstateRole::NORMAL) {
-        if (fScriptChecks) {
-            Assume(!script_check_reason.empty());
+    if (script_check_reason != m_prev_script_check_reason && GetRole() == ChainstateRole::NORMAL) {
+        if (script_check_reason) {
             LogInfo("Enabling script verification at block #%d (%s): %s.",
                     pindex->nHeight, block_hash.ToString(), script_check_reason);
         } else {
             LogInfo("Disabling script verification at block #%d (%s).",
                     pindex->nHeight, block_hash.ToString());
         }
-        m_prev_script_checks_logged = fScriptChecks;
+        m_prev_script_check_reason = script_check_reason;
     }
 
     CBlockUndo blockundo;
@@ -2593,7 +2590,7 @@ bool Chainstate::ConnectBlock(const CBlock& block, BlockValidationState& state,
     // doesn't invalidate pointers into the vector, and keep txsdata in scope
     // for as long as `control`.
     std::optional<CCheckQueueControl<CScriptCheck>> control;
-    if (auto& queue = m_chainman.GetCheckQueue(); queue.HasThreads() && fScriptChecks) control.emplace(queue);
+    if (auto& queue = m_chainman.GetCheckQueue(); queue.HasThreads() && script_check_reason) control.emplace(queue);
 
     std::vector<PrecomputedTransactionData> txsdata(block.vtx.size());
 
@@ -2652,7 +2649,7 @@ bool Chainstate::ConnectBlock(const CBlock& block, BlockValidationState& state,
             break;
         }
 
-        if (!tx.IsCoinBase() && fScriptChecks)
+        if (!tx.IsCoinBase() && script_check_reason)
         {
             bool fCacheResults = fJustCheck; /* Don't cache results if we're actually connecting blocks (still consult the cache, though) */
             bool tx_ok;
diff --git a/src/validation.h b/src/validation.h
index fea11c5515..6e426a4660 100644
--- a/src/validation.h
+++ b/src/validation.h
@@ -560,7 +560,7 @@ protected:
     //! Cached result of LookupBlockIndex(*m_from_snapshot_blockhash)
     mutable const CBlockIndex* m_cached_snapshot_base GUARDED_BY(::cs_main){nullptr};
 
-    std::optional<bool> m_prev_script_checks_logged GUARDED_BY(::cs_main){};
+    std::optional<const char*> m_prev_script_check_reason GUARDED_BY(::cs_main){};
 
 public:
     //! Reference to a BlockManager instance which itself is shared across all
```
</details>

l0rinc replied:
I don't think it makes much difference, but looks like you do, so I have changed it to `const char*`, let me know what you think.

hodlinator reviewed (commented):
Reviewed f6951cb74bba60fedb2e0c041038561ba2583594

hodlinator commented on test/functional/feature_assumevalid.py:
```diff
@@ -140,35 +141,68 @@ def run_test(self):
-        with self.nodes[1].assert_debug_log(expected_msgs=['Disabling signature validations at block #1', 'Enabling signature validations at block #103']):
+        with self.nodes[1].assert_debug_log(expected_msgs=['Disabling signature validations at block #1',  # AssumeValid::SKIPPED
+                                                           'Enabling signature validations at block #103',
+                                                           'block not part of assumevalid chain']):  # AssumeValid::CHECKED_NOT_UNDER_ASSUMEVALID
```
Yes, I would prefer having a separate reason message for when we have passed beyond the assumevalid height since it's so fundamental to assumevalid, and so different from being at a height below assumevalid on a forked chain.

Agree it probably belongs in a separate commit, but still think it would be good to have in this PR.
Your first commit that currently introduces `"block not part of assumevalid chain"` should probably state something more neutral like `"block is not an ancestor of the assumevalid block"`.
Then with the separate commit:
```diff
+        } else if (it->second.nHeight < pindex->nHeight) {
+            script_check_reason = "block height exceeds assumevalid height";
         } else if (it->second.GetAncestor(pindex->nHeight) != pindex) {
-            script_check_reason = "block is not an ancestor of the assumevalid block";
+            script_check_reason = "block not part of assumevalid chain";
```

l0rinc commented on src/validation.cpp:
```diff
@@ -2423,35 +2423,45 @@ bool Chainstate::ConnectBlock(const CBlock& block, BlockValidationState& state,
     }
 
-    bool fScriptChecks = true;
+    auto fScriptChecks{true};
```
I think this is trivial to review.

l0rinc commented on test/functional/feature_assumevalid.py:
```diff
@@ -137,38 +131,98 @@ def run_test(self):
+            second_chain.append(block)
+            second_chain_tip, second_chain_time, second_chain_height = block.hash_int, second_chain_time + 1, second_chain_height + 1
+
+        p2p4 = self.nodes[3].add_p2p_connection(BaseNode())
```
Fixed name

l0rinc commented on test/functional/feature_assumevalid.py:
```diff
@@ -129,19 +131,23 @@ def run_test(self):
             self.block_time += 1
             height += 1
 
-        # Start node1 and node2 with assumevalid so they accept a block with a bad signature.
-        self.start_node(1, extra_args=["-assumevalid=" + block102.hash_hex])
-        self.start_node(2, extra_args=["-assumevalid=" + block102.hash_hex])
```
No problem, I put them back to the beginning

l0rinc committed:
log: split assumevalid ancestry-failure-reason message

When the assumevalid ancestry check fails, log a precise reason:
- "block height above assumevalid height" if the block is above the assumevalid block (the default reason)
- "block not in of assumevalid chain" otherwise

The new split was added under the existing condition to simplify conceptually that the two cases are related.
It could still be useful to know when the block is just above the assumevalid block or when it's not even on the same chain.

Update the functional test to assert the new reason strings. No behavior change.

Co-authored-by: Hodlinator <172445034+hodlinator@users.noreply.github.com>

l0rinc reviewed (commented):
Thanks for the comments, adjusted the test and code to differentiate between cases when the block is after the assumevalid block from when it's not even on the same chain.
Also adjusted the style of the test to have more self-contained blocks separated by the `with` blocks, unified the test string styles, updated old commit messages.
Pushed the [change](https://github.com/bitcoin/bitcoin/compare/f66036d47e221bed7fe94c517631be920da66d96..fb5405e43acf6d0e6ce74643f6372410ef8cc41d) and the rebase separately.

hodlinator commented on test/functional/feature_assumevalid.py:
```diff
@@ -170,5 +179,21 @@ def run_test(self):
+        p2p3 = self.nodes[3].add_p2p_connection(BaseNode())
+        p2p3.send_header_for_blocks(self.blocks[0:200])
+        p2p3.send_without_ping(msg_block(self.blocks[0]))
+        self.wait_until(lambda: self.nodes[3].getblockcount())
```
(Thanks for changing to `self.wait_until(lambda: self.nodes[3].getblockcount() == 1)`).

hodlinator commented on test/functional/feature_assumevalid.py:
```diff
@@ -28,6 +27,12 @@
+      competing headers-only branch so block #1 is not on the best header chain.
+    - node4 has -assumevalid set to the hash of block 102. Submit an alternative
+      block #1 that is not part of the assumevalid chain.
+    - node5 has no -assumevalid parameter. Reindex to deterministically hit
+      "assumevalid hash not in headers" and "below minimum chainwork".
```
nit:
```suggestion
    - node5 starts with no -assumevalid parameter. Reindex to deterministically hit
      "assumevalid hash not in headers" and "below minimum chainwork".
```

l0rinc replied:
Done

hodlinator commented on test/functional/feature_assumevalid.py:
```diff
@@ -136,38 +141,117 @@ def run_test(self):
-        p2p1.send_header_for_blocks(self.blocks[0:2000])
-        p2p1.send_header_for_blocks(self.blocks[2000:])
-        with self.nodes[1].assert_debug_log(expected_msgs=['Disabling signature validations at block #1', 'Enabling signature validations at block #103']):
+        with self.nodes[0].assert_debug_log(expected_msgs=[
+            f"Enabling script verification at block #1 ({block_1_hash}): assumevalid=0 (always verify).",
+            "Block validation error: block-script-verify-flag-failed",
+        ]):
+            p2p0 = self.nodes[0].add_p2p_connection(BaseNode())
+
+            p2p0.send_header_for_blocks(self.blocks[0:2000])
+            p2p0.send_header_for_blocks(self.blocks[2000:])
+
+            self.send_blocks_until_disconnected(p2p0)
+            self.wait_until(lambda: self.nodes[0].getblockcount() >= COINBASE_MATURITY + 1)
+            assert_equal(self.nodes[0].getblockcount(), COINBASE_MATURITY + 1)
```
nit: Had a back & forth out-of-band about increasing the `with`-block scopes. While it makes the code layout nicer, it makes for slightly sloppier tests. But initializing a p2p connection and then doing some asserts at the end of the block doesn't seem too bad, so this is nit-level.

hodlinator reviewed (commented):
Reviewed 38025edd1978670bfa3c32cf518acd0b01644167

l0rinc commented on src/validation.h:
```diff
@@ -560,7 +560,7 @@ class Chainstate
     mutable const CBlockIndex* m_cached_snapshot_base GUARDED_BY(::cs_main){nullptr};
 
-    std::optional<bool> m_prev_script_checks_logged GUARDED_BY(::cs_main){};
+    std::optional<std::string> m_last_script_check_reason_logged GUARDED_BY(::cs_main){""}; // init with foreign value
```
> Don't see much of a point to change m_prev_script_checks_logged types twice

I deliberately separated "enabling" logs changes from enabling reasons to show the changes in small steps, reflected in the tests. They're fundamentally different changes that are reflected in separate test and behavior changes, do you feel strongly about merging them?

hodlinator replied:
Curious what other reviewers think, but not a blocker.

hodlinator reviewed (approved):
re-ACK 010cf81407c0df8de766fd2a116463d180f25f33

Eunovo reviewed (commented):
Left a few comments

w0xlt reviewed (commented):
reACK https://github.com/bitcoin/bitcoin/pull/33336/commits/010cf81407c0df8de766fd2a116463d180f25f33

ajtowns reviewed (commented):
Some possible nits

Eunovo reviewed (commented):
ACK https://github.com/bitcoin/bitcoin/pull/33336/commits/010cf81407c0df8de766fd2a116463d180f25f33

andrewtoth reviewed (approved):
Approach ACK.

The way this PR untangles the assumevalid conditions and explicitly logs them is great for readability and debugging. I'm not sure this should go into v30 though since it does touch a lot of consensus code.

The PR title understates quite a bit what this PR is doing. I assume since it has evolved since being initially proposed. I think it can be updated to better detail what this is doing.

l0rinc commented on test/functional/feature_assumevalid.py:
```diff
@@ -136,40 +142,109 @@ def run_test(self):
+        p2p5 = self.nodes[5].add_p2p_connection(BaseNode())
+        p2p5.send_header_for_blocks(self.blocks[0:200])
+        p2p5.send_without_ping(msg_block(self.blocks[0]))
+        self.wait_until(lambda: self.nodes[5].getblockcount())
```
Wouldn't that cause a race condition if the check misses it?

hodlinator replied:
But you've only sent it one block?

l0rinc replied:
Right, so what's the advantage of `getblockcount() == 1` compared to "non-zero"?

hodlinator replied:
Makes the int -> bool coercion explicit to the reader, see https://github.com/bitcoin/bitcoin/pull/33336#discussion_r2362266224

hodlinator commented on src/validation.cpp:
```diff
@@ -2433,15 +2433,15 @@ bool Chainstate::ConnectBlock(const CBlock& block, BlockValidationState& state,
+            script_check_reason = "block not part of assumevalid chain";
         } else if (m_chainman.m_best_header->GetAncestor(pindex->nHeight) != pindex) {
-            // TODO
+            script_check_reason = "block not on best header chain";
```
Blocks are "in the chain" to me, like metal links being in a chain. Maybe on-chain tx is more in contrast to off-chain tx, but a tx is in a block, in a chain.

Not important enough that I reacted upon first review, but would definitely suggest changing if you re-touch.

l0rinc replied:
Fixed, thanks

l0rinc commented on src/validation.cpp:
```diff
@@ -2563,9 +2572,15 @@ bool Chainstate::ConnectBlock(const CBlock& block, BlockValidationState& state,
+            LogInfo("Disabling script verification at block #%d (%s).",
+                    pindex->nHeight, block_hash.ToString());
+        }
+        m_last_script_check_reason_logged = script_check_reason;
```
I think I understood that, but that will be a behavior change, since currently after `Enabling... too recent...` we're still getting a log for `Enabling ... block height above...` while if we only check empty vs non-empty we'd only be logging `Disabled` -> `Enabled` (or vice versa) but not slightly more confusing cases like the one above. Given that the logging was added because even we got confused by this, I thought we should log every state change.

andrewtoth replied:
I don't understand why you think it's a behavior change. It has nothing to do with logging. Add this on line 2585: `const bool fScriptChecks = (script_check_reason != nullptr);`
Remove changes on lines 2594, 2653.

l0rinc commented on test/functional/feature_assumevalid.py:
```diff
@@ -27,6 +27,12 @@
+      competing headers-only branch so block #1 is not on the best header chain.
+    - node4 has -assumevalid set to the hash of block 102. Submit an alternative
+      block #1 that is not part of the assumevalid chain.
+    - node5 starts with no -assumevalid parameter. Reindex to deterministically hit
```
Removed

l0rinc commented on test/functional/feature_assumevalid.py:
```diff
@@ -136,40 +142,109 @@ def run_test(self):
+
+
+        # nodes[5]
+        # Reindex to deterministically hit specific assumevalid gates (no races with header downloads/chainwork during startup).
```
Removed

l0rinc commented on src/validation.cpp:
```diff
@@ -2433,15 +2433,15 @@ bool Chainstate::ConnectBlock(const CBlock& block, BlockValidationState& state,
+            script_check_reason = "assumevalid hash not in headers";
         } else if (it->second.GetAncestor(pindex->nHeight) != pindex) {
-            // TODO
+            script_check_reason = "block not part of assumevalid chain";
```
Good point, thanks

hodlinator reviewed (approved):
re-ACK 45bd8914658a675d00aa9c83373d6903a8a9ece8

Just fixed some nits since previous review.

yuvicc reviewed (commented):
ACK 45bd8914658a675d00aa9c83373d6903a8a9ece8

Reviewed all commits.

andrewtoth reviewed (approved):
ACK 45bd8914658a675d00aa9c83373d6903a8a9ece8

ajtowns commented on src/validation.cpp:
```diff
@@ -2423,34 +2423,43 @@ bool Chainstate::ConnectBlock(const CBlock& block, BlockValidationState& state,
 
-    bool fScriptChecks = true;
-    if (!m_chainman.AssumedValidBlock().IsNull()) {
+    const char* script_check_reason;
```
nit: Would be better to initialise to `nullptr`, IMO. Sanitizers should catch and warn for initialisation being missed in tested code paths though.

ajtowns commented on src/validation.cpp:
```diff
@@ -2423,34 +2423,43 @@ bool Chainstate::ConnectBlock(const CBlock& block, BlockValidationState& state,
+        if (it == m_blockman.m_block_index.end()) {
+            script_check_reason = "assumevalid hash not in headers";
+        } else if (it->second.GetAncestor(pindex->nHeight) != pindex) {
+            script_check_reason = (pindex->nHeight > it->second.nHeight) ? "block height above assumevalid height" : "block not in assumevalid chain";
```
nit: Better as an `if` than ` .. ? .. : ..` IMO.

ajtowns reviewed (commented):
ACK 45bd8914658a675d00aa9c83373d6903a8a9ece8

nits don't need fixing
